// bs-broken-links-checker/lib/reporters/html.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _base = require('./base');

var _base2 = _interopRequireDefault(_base);

/**
 * @class ReporterHtml
 * @desc Class for creating html report
 */

var ReporterHtml = (function (_ReporterBase) {
    _inherits(ReporterHtml, _ReporterBase);

    function ReporterHtml() {
        _classCallCheck(this, ReporterHtml);

        _get(Object.getPrototypeOf(ReporterHtml.prototype), 'constructor', this).apply(this, arguments);
    }

    _createClass(ReporterHtml, [{
        key: 'createReport',
        value: function createReport(configurationName, statistic, options) {
            this._logger.info('create html report');
            var report = {
                name: options.url,
                date: (0, _moment2['default'])().format('DD-MM-YYYY:hh:mm:ss'),
                internalCount: statistic.getInternalCount(),
                externalCount: statistic.getExternalCount(),
                totalCount: statistic.getAllCount(),
                brokenCount: statistic.getBrokenCount(),
                broken: statistic.getBroken().getAll(),
                options: options
            },
                htmlTemplate = _fs2['default'].readFileSync(_path2['default'].resolve(__dirname, '../../src/assets/report.html'), { encoding: 'utf-8' }),
                compiled = _lodash2['default'].template(htmlTemplate);

            return this.saveReportFile(configurationName, 'html', compiled({ report: report }), report.date);
        }
    }]);

    return ReporterHtml;
})(_base2['default']);

exports['default'] = ReporterHtml;
module.exports = exports['default'];
