// bs-broken-links-checker/lib/reporters/json.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _base = require('./base');

var _base2 = _interopRequireDefault(_base);

/**
 * @class ReporterJson
 * @desc Class for creating json report
 */

var ReporterJson = (function (_ReporterBase) {
    _inherits(ReporterJson, _ReporterBase);

    function ReporterJson() {
        _classCallCheck(this, ReporterJson);

        _get(Object.getPrototypeOf(ReporterJson.prototype), 'constructor', this).apply(this, arguments);
    }

    _createClass(ReporterJson, [{
        key: 'createReport',
        value: function createReport(configurationName, statistic) {
            this._logger.info('create json report');
            var report = {
                date: (0, _moment2['default'])().format('DD-MM-YYYY:hh:mm:ss'),
                internalCount: statistic.getInternalCount(),
                externalCount: statistic.getExternalCount(),
                totalCount: statistic.getAllCount(),
                brokenCount: statistic.getBrokenCount(),
                broken: statistic.getBroken().getAll()
            };

            return this.saveReportFile(configurationName, 'json', JSON.stringify(report, null, 4), report.date);
        }
    }]);

    return ReporterJson;
})(_base2['default']);

exports['default'] = ReporterJson;
module.exports = exports['default'];
